File: HOT100/12.栈/155.最小栈.cpp
/*
 * @lc app=leetcode.cn id=155 lang=cpp
 *
 * [155] 最小栈
 */
#include "tools.h"

// @lc code=start
class MinStack {
    stack<pair<int, int>> stk;
public:
    MinStack() {
        stk.emplace(0, INT_MAX);
    }
    
    void push(int val) {
        stk.emplace(val, min(getMin(), val));
    }
    
    void pop() {
        stk.pop();
    }
    
    int top() {
        return stk.top().first;
    }
    
    int getMin() {
        return stk.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
// @lc code=end

File: HOT100/HOT100p1/12.栈/155.最小栈.py
#
# @lc app=leetcode.cn id=155 lang=python3
#
# [155] 最小栈
#
from mytools import *
# @lc code=start
class MinStack:

    def __init__(self):
        self.stk = [(0, float('inf'))]

    def push(self, val: int) -> None:
        self.stk.append((val, min(self.getMin(), val)))

    def pop(self) -> None:
        self.stk.pop()

    def top(self) -> int:
        return self.stk[-1][0]

    def getMin(self) -> int:
        return self.stk[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
# @lc code=end

